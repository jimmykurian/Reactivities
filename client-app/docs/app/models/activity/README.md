[**Client App Documentation v0.0.0**](../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../README.md) / app/models/activity

# app/models/activity

## Index

### Interfaces

- [Activity](interfaces/Activity.md)
