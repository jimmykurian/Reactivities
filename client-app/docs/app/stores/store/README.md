[**Client App Documentation v0.0.0**](../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../README.md) / app/stores/store

# app/stores/store

## Index

### Interfaces

- [Store](interfaces/Store.md)

### Variables

- [StoreContext](variables/StoreContext.md)
- [store](variables/store.md)

### Functions

- [useStore](functions/useStore.md)
