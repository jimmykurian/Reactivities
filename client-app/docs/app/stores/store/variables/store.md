[**Client App Documentation v0.0.0**](../../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../../README.md) / [app/stores/store](../README.md) / store

# Variable: store

> `const` **store**: [`Store`](../interfaces/Store.md)

## Constant

store

## Description

The store instance containing all the application's stores.

## Source

[src/app/stores/store.ts:25](https://github.com/jimmykurian/Reactivities/blob/5670213c338d2af4595c1ce87506b162bb3099b4/client-app/src/app/stores/store.ts#L25)
