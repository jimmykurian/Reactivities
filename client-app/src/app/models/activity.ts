/**
 * @author Jimmy Kurian
 * @name Activity
 */

/**
 * The Activity interface represents the structure of an activity object in the Reactivities application.
 * It includes various properties such as id, title, date, description, category, city, and venue.
 *
 * @interface
 *
 * @property {string} id - The unique identifier for the activity.
 * @property {string} title - The title of the activity.
 * @property {string} date - The date of the activity in ISO string format.
 * @property {string} description - A brief description of the activity.
 * @property {string} category - The category to which the activity belongs.
 * @property {string} city - The city where the activity takes place.
 * @property {string} venue - The specific venue of the activity within the city.
 */
export interface Activity {
  id: string;
  title: string;
  date: string;
  description: string;
  category: string;
  city: string;
  venue: string;
}
