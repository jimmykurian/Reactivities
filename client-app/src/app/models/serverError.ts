/**
 * @file ServerError.ts
 * @author Jimmy Kurian
 * @fileoverview This file defines the `ServerError` interface, which models the structure of server error responses in the Reactivities application.
 */

/**
 * The ServerError interface represents the structure of an error returned from the server.
 * It includes the HTTP status code, an error message, and optional detailed information about the error.
 *
 * @interface ServerError
 *
 * @property {number} statusCode - The HTTP status code associated with the error.
 * @property {string} message - A brief message describing the error.
 * @property {string} details - Additional details or stack trace information about the error (optional).
 */
export interface ServerError {
  statusCode: number;
  message: string;
  details: string;
}
