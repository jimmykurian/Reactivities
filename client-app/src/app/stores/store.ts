/**
 * @author Jimmy Kurian
 * @name Store
 */

import { createContext, useContext } from 'react';
import ActivityStore from './activityStore';

/**
 * Interface representing the structure of the store.
 *
 * @interface Store
 * @property {ActivityStore} activityStore - Instance of ActivityStore for managing activity-related state.
 */
interface Store {
  activityStore: ActivityStore;
}

/**
 * The store instance containing all the application's stores.
 *
 * @constant
 * @type {Store}
 */
export const store: Store = {
  activityStore: new ActivityStore(),
};

/**
 * The React context for the store.
 *
 * @constant
 * @type {React.Context<Store>}
 */
export const StoreContext = createContext(store);

/**
 * Custom hook to use the store in React components.
 *
 * @function
 * @returns {Store} The store instance.
 *
 * @example
 * ```tsx
 * import { useStore } from './store';
 *
 * const MyComponent = () => {
 *   const { activityStore } = useStore();
 *   return <div>{activityStore.title}</div>;
 * };
 * ```
 */
export function useStore(): Store {
  return useContext(StoreContext);
}
