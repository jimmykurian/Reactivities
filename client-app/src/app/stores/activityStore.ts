/**
 * @file activityStore.tsx
 * @author Jimmy Kurian
 * @fileoverview This file contains the ActivityStore class which manages the state of activities using MobX.
 */

import { makeAutoObservable, runInAction } from 'mobx';
import { Activity } from '../models/activity';
import agent from '../api/agent';
import { v4 as uuid } from 'uuid';

/**
 * @class ActivityStore
 * @classdesc This class represents the MobX store for managing activity-related state.
 * It initializes the observables and configures MobX for state management.
 *
 * @example
 * ```tsx
 * import ActivityStore from './ActivityStore';
 *
 * const activityStore = new ActivityStore();
 * activityStore.loadActivities();
 * ```
 */
export default class ActivityStore {
  /** @type {Map<string, Activity>} */
  activityRegistry: Map<string, Activity> = new Map<string, Activity>();

  /** @type {Activity | undefined} */
  selectedActivity: Activity | undefined = undefined;

  /** @type {boolean} */
  editMode: boolean = false;

  /** @type {boolean} */
  loading: boolean = false;

  /** @type {boolean} */
  loadingInitial: boolean = true;

  /**
   * @constructor ActivityStore
   * @description Creates an instance of ActivityStore.
   * Initializes the observables using MobX's makeAutoObservable.
   */
  constructor() {
    makeAutoObservable(this);
  }

  /**
   * @getter activitiesByDate
   * @description Gets the activities sorted by date.
   * @returns {Activity[]}
   */
  get activitiesByDate(): Activity[] {
    return Array.from(this.activityRegistry.values()).sort(
      (a, b) => Date.parse(a.date) - Date.parse(b.date),
    );
  }

  /**
   * @async
   * @function loadActivities
   * @description Loads activities from the API and sets the activities state.
   * @returns {Promise<void>}
   */
  loadActivities = async (): Promise<void> => {
    try {
      const activities = await agent.Activities.list();
      runInAction(() => {
        activities.forEach((activity: Activity) => {
          activity.date = activity.date.split('T')[0];
          this.activityRegistry.set(activity.id, activity);
        });
        this.setLoadingInitial(false);
      });
    } catch (error) {
      console.log(error);
      runInAction(() => {
        this.setLoadingInitial(false);
      });
    }
  };

  /**.
   * @function setLoadingInitial
   * @description Sets the loading initial state
   * @param {boolean} state - The loading state to set.
   */
  setLoadingInitial = (state: boolean): void => {
    this.loadingInitial = state;
  };

  /**
   * @function selectActivity
   * @description Selects an activity by its ID.
   * @param {string} id - The ID of the activity to select.
   */
  selectActivity = (id: string): void => {
    this.selectedActivity = this.activityRegistry.get(id);
  };

  /**
   * @function cancelSelectedActivity
   * @description Cancels the selection of the currently selected activity.
   */
  cancelSelectedActivity = (): void => {
    this.selectedActivity = undefined;
  };

  /**
   * @function openForm
   * @description Opens the form for creating or editing an activity.
   * @param {string} [id] - The ID of the activity to edit, if any.
   */
  openForm = (id?: string): void => {
    id ? this.selectActivity(id) : this.cancelSelectedActivity();
    this.editMode = true;
  };

  /**
   * @function closeForm
   * @description Closes the form for creating or editing an activity.
   */
  closeForm = (): void => {
    this.editMode = false;
  };

  /**
   * @async
   * @function createActivity
   * @description Creates a new activity.
   * @param {Activity} activity - The activity to create.
   * @returns {Promise<void>}
   */
  createActivity = async (activity: Activity): Promise<void> => {
    this.loading = true;
    activity.id = uuid();
    try {
      await agent.Activities.create(activity);
      runInAction(() => {
        this.activityRegistry.set(activity.id, activity);
        this.selectedActivity = activity;
        this.editMode = false;
        this.loading = false;
      });
    } catch (error) {
      console.log(error);
      runInAction(() => {
        this.loading = false;
      });
    }
  };

  /**
   * @async
   * @function updateActivity
   * @description Updates an existing activity.
   * @param {Activity} activity - The activity to update.
   * @returns {Promise<void>}
   */
  updateActivity = async (activity: Activity): Promise<void> => {
    this.loading = true;
    try {
      await agent.Activities.update(activity);
      runInAction(() => {
        this.activityRegistry.set(activity.id, activity);
        this.selectedActivity = activity;
        this.editMode = false;
        this.loading = false;
      });
    } catch (error) {
      console.log(error);
      runInAction(() => {
        this.loading = false;
      });
    }
  };

  /**
   * @async
   * @function deleteActivity
   * @description Deletes an activity by its ID.
   * @param {string} id - The ID of the activity to delete.
   * @returns {Promise<void>}
   */
  deleteActivity = async (id: string): Promise<void> => {
    this.loading = true;
    try {
      await agent.Activities.delete(id);
      runInAction(() => {
        this.activityRegistry.delete(id);
        if (this.selectedActivity?.id === id) this.cancelSelectedActivity();
        this.loading = false;
      });
    } catch (error) {
      console.log(error);
      runInAction(() => {
        this.loading = false;
      });
    }
  };
}
