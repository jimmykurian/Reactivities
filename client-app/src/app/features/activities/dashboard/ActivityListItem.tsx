/**
 * @file ActivityListItem.tsx
 * @author Jimmy Kurian
 * @fileoverview This file contains the ActivityListItem component.
 */

import { Link } from 'react-router-dom';
import { Item, Button, Segment, Icon } from 'semantic-ui-react';
import { Activity } from '../../../models/activity';
import { useState, SyntheticEvent } from 'react';
import { useStore } from '../../../stores/store';

/**
 * @interface Props
 * @description The props object for the ActivityListItem component.
 * @property {Activity} activity - The activity to display.
 */
export interface Props {
  activity: Activity;
}

/**
 * @component ActivityListItem
 * @description A functional component that renders an individual activity item.
 *
 * @param {Props} props - The props object containing the activity data.
 * @returns {JSX.Element} The JSX element representing an individual activity item.
 *
 * @remarks
 * This component uses the Semantic UI React library to create a styled item layout
 * for displaying individual activity details such as title, date, description, city, and venue.
 * It includes buttons for viewing and deleting the activity. The deletion process is handled
 * by the `handleActivityDelete` function which interacts with the MobX store's deleteActivity function.
 *
 * @example
 * Here is an example of how to use the ActivityListItem component:
 * ```tsx
 * import React from 'react';
 * import { Activity } from '../../../models/activity';
 * import ActivityListItem from './ActivityListItem';
 *
 * const activity: Activity = {
 *   id: '1',
 *   title: 'Morning Run',
 *   date: '2024-06-12',
 *   description: 'A quick morning run around the park.',
 *   category: 'exercise',
 *   city: 'New York',
 *   venue: 'Central Park'
 * };
 *
 * <ActivityListItem activity={activity} />
 * ```
 */
export default function ActivityListItem({ activity }: Props): JSX.Element {
  const { activityStore } = useStore();
  const { deleteActivity, loading } = activityStore;
  const [target, setTarget] = useState('');

  /**
   * @function handleActivityDelete
   * @description Handles the deletion of an activity.
   * @param {SyntheticEvent<HTMLButtonElement>} e - The synthetic event triggered by clicking the delete button.
   * @param {string} id - The ID of the activity to delete.
   */
  function handleActivityDelete(
    e: SyntheticEvent<HTMLButtonElement>,
    id: string,
  ): void {
    setTarget(e.currentTarget.name);
    deleteActivity(id);
  }

  return (
    <Segment.Group>
      <Segment>
        <Item.Group>
          <Item>
            <Item.Image size="tiny" circular src="/assets/user.png" />
            <Item.Content>
              <Item.Header as={Link} to={`/activities/${activity.id}`}>
                {activity.title}
              </Item.Header>
              <Item.Description>Hosted by Bob</Item.Description>
            </Item.Content>
          </Item>
        </Item.Group>
      </Segment>
      <Segment>
        <span>
          <Icon name="clock" /> {activity.date}
          <Icon name="marker" /> {activity.venue}
        </span>
      </Segment>
      <Segment secondary>Attendees will go here</Segment>
      <Segment clearing>
        <span>{activity.description}</span>
        <Button
          as={Link}
          to={`/activities/${activity.id}`}
          color="teal"
          floated="right"
          content="View"
        />
        <Button
          name={activity.id}
          loading={loading && target === activity.id}
          onClick={(e) => handleActivityDelete(e, activity.id)}
          color="red"
          floated="right"
          content="Delete"
        />
      </Segment>
    </Segment.Group>
  );
}
