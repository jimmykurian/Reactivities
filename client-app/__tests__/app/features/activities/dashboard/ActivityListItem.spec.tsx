import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import ActivityListItem from '../../../../../src/app/features/activities/dashboard/ActivityListItem';
import { faker } from '@faker-js/faker';
import { Activity } from '../../../../../src/app/models/activity';
import { useStore } from '../../../../../src/app/stores/store';
import { MemoryRouter } from 'react-router-dom';

// Mock the store context
jest.mock('../../../../../src/app/stores/store', () => ({
  useStore: jest.fn(),
}));

const generateMockActivity = (): Activity => ({
  id: faker.string.uuid(),
  title: faker.lorem.words(3),
  date: faker.date.future().toISOString(),
  description: faker.lorem.sentence(),
  category: faker.lorem.word(),
  city: faker.location.city(),
  venue: faker.location.street(),
});

// Static data for snapshot test
const staticActivity: Activity = {
  id: '1',
  title: 'Static Activity',
  date: '2024-07-28T00:00:00.000Z',
  description: 'This is a static activity description.',
  category: 'Static Category',
  city: 'Static City',
  venue: 'Static Venue',
};

describe('ActivityListItem', () => {
  const deleteActivity = jest.fn();
  const loading = false;

  const mockStore = {
    activityStore: {
      deleteActivity,
      loading,
    },
  };

  beforeEach(() => {
    (useStore as jest.Mock).mockReturnValue(mockStore);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders ActivityListItem component', () => {
    // Arrange
    const mockActivity = generateMockActivity();
    render(
      <MemoryRouter>
        <ActivityListItem activity={mockActivity} />
      </MemoryRouter>,
    );

    // Act & Assert
    expect(screen.getByText(mockActivity.title)).toBeInTheDocument();
    expect(screen.getByText(mockActivity.date)).toBeInTheDocument();
    expect(screen.getByText(mockActivity.description)).toBeInTheDocument();
    expect(
      screen.getByText(`${mockActivity.city}, ${mockActivity.venue}`),
    ).toBeInTheDocument();
    expect(screen.getByText(mockActivity.category)).toBeInTheDocument();
  });

  test('calls deleteActivity when delete button is clicked', async () => {
    // Arrange
    const mockActivity = generateMockActivity();
    render(
      <MemoryRouter>
        <ActivityListItem activity={mockActivity} />
      </MemoryRouter>,
    );

    const deleteButton = screen.getByRole('button', { name: /delete/i });

    // Act
    fireEvent.click(deleteButton);

    // Assert
    await waitFor(() => {
      expect(deleteActivity).toHaveBeenCalledWith(mockActivity.id);
    });
  });

  test('displays loading state when delete button is clicked', async () => {
    // Arrange
    const mockActivity = generateMockActivity();
    mockStore.activityStore.loading = true;
    (useStore as jest.Mock).mockReturnValue(mockStore);

    render(
      <MemoryRouter>
        <ActivityListItem activity={mockActivity} />
      </MemoryRouter>,
    );

    const deleteButton = screen.getByRole('button', { name: /delete/i });

    // Act
    fireEvent.click(deleteButton);

    // Assert
    await waitFor(() => {
      expect(deleteButton).toHaveClass('loading');
    });
  });

  test('matches snapshot', () => {
    // Arrange & Act
    const { asFragment } = render(
      <MemoryRouter>
        <ActivityListItem activity={staticActivity} />
      </MemoryRouter>,
    );

    // Assert
    expect(asFragment()).toMatchSnapshot();
  });
});
