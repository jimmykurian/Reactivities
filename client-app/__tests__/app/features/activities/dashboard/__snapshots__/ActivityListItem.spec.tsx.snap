// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityListItem matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui segments"
  >
    <div
      class="ui segment"
    >
      <div
        class="ui items"
      >
        <div
          class="item"
        >
          <div
            class="ui tiny circular image"
          >
            <img
              src="/assets/user.png"
            />
          </div>
          <div
            class="content"
          >
            <a
              class="header"
              href="/activities/1"
            >
              Static Activity
            </a>
            <div
              class="description"
            >
              Hosted by Bob
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ui segment"
    >
      <span>
        <i
          aria-hidden="true"
          class="clock icon"
        />
         2024-07-28
        <i
          aria-hidden="true"
          class="marker icon"
        />
         Static Venue
      </span>
    </div>
    <div
      class="ui secondary segment"
    >
      Attendees will go here
    </div>
    <div
      class="ui clearing segment"
    >
      <span>
        This is a static activity description.
      </span>
      <a
        class="ui teal right floated button"
        href="/activities/1"
        role="button"
      >
        View
      </a>
      <button
        class="ui red right floated button"
        name="1"
      >
        Delete
      </button>
    </div>
  </div>
</DocumentFragment>
`;
